src/Core/Configurations/RuntimeConfigValidator.cs

```diff
@@ -830,11 +833,107 @@ public void ValidateRelationshipsInConfig(RuntimeConfig runtimeConfig, IMetadata
+                if (string.IsNullOrWhiteSpace(relationship.LinkingObject))
+                {
+                    // Validation to ensure that source and target fields are both null or both not null.
+                    if (relationship.SourceFields is null ^ relationship.TargetFields is null)
+                    {
+                        HandleOrRecordException(new DataApiBuilderException(
+                            message: $"Entity: {entityName} has a relationship: {relationshipName}, which has source and target fields " +
+                                $"where one is null and the other is not.",
+                            statusCode: HttpStatusCode.ServiceUnavailable,
+                            subStatusCode: DataApiBuilderException.SubStatusCodes.ConfigValidationError));
+                    }
+
+                    // In one-one, one-many, or many-one relationships when both source and target are non null their size must match.
+                    if ((relationship.SourceFields is not null && relationship.TargetFields is not null) &&
+                        relationship.SourceFields.Length != relationship.TargetFields.Length)
+                    {
+                        HandleOrRecordException(new DataApiBuilderException(
+                            message: $"Entity: {entityName} has a relationship: {relationshipName}, which has {relationship.SourceFields.Length} source fields defined, " +
+                                $"but {relationship.TargetFields.Length} target fields defined.",
+                            statusCode: HttpStatusCode.ServiceUnavailable,
+                            subStatusCode: DataApiBuilderException.SubStatusCodes.ConfigValidationError));
+                    }
+                }
+
+                // In all kinds of relationships, if sourceFields are included they must be valid columns in the backend.
+                if (relationship.SourceFields is not null)
+                {
+                    foreach (string sourceField in relationship.SourceFields)
+                    {
+                        if (!sqlMetadataProvider.TryGetBackingColumn(entityName, sourceField, out _))
+                        {
+                            invalidColumns.Add(sourceField);
+                        }
+
+                        if (invalidColumns.Count > 0)
+                        {
+                            HandleOrRecordException(new DataApiBuilderException(
+                                message: $"Entity: {entityName} has a relationship: {relationshipName} with source fields: {string.Join(",", invalidColumns)} that " +
+                                    $"do not exist as columns in {entityName}.",
+                                statusCode: HttpStatusCode.ServiceUnavailable,
+                                subStatusCode: DataApiBuilderException.SubStatusCodes.ConfigValidationError));
+                            invalidColumns.Clear();
+                        }
```

Comment:
When you clear `invalidColumns` after each occurence of detecting an invalid column, you'll never have `invalidColumns.Count` be greater than 1. So there doesn't seem to be a benefit to `string.Join(",", invalidColumns)`. Did you mean to have this if block be outside the for loop? 

If you intended the `if` block to be outside the for loop, you can further simplify the code checking for valid source and valid target fields.

Reply:
Yes, this was supposed to be outside the loop, per our discussion, ive moved it all a helper function.